MdeModulePkg/Library/DeviceStateLib/DeviceStateLib.c

```diff
@@ -13,30 +13,129 @@ MU_CHANGE: new file
 #include <Library/DeviceStateLib.h>
 #include <Library/DebugLib.h>
 
+CHAR8  *DEVICE_STATE_STRINGS[6] = {
+  DEVICE_STATE_SECUREBOOT_OFF_STRING,
+  DEVICE_STATE_MANUFACTURING_MODE_STRING,
+  DEVICE_STATE_DEVELOPMENT_BUILD_ENABLED_STRING,
+  DEVICE_STATE_SOURCE_DEBUG_ENABLED_STRING,
+  DEVICE_STATE_UNDEFINED_STRING,
+  DEVICE_STATE_UNIT_TEST_MODE_STRING,
+};
+
 /**
-Function to Get current device state
-@retval the current DEVICE_STATE
-**/
+ * Get String that represents the device's current insecure states, using the platform defined set of
+ * insecure device states. This can be used to measure insecure device states into the TPM or perform
+ * other required platform actions when the device enters an insecure state.
+ *
+ * @param[out] Buffer - Buffer to store the string
+ * @param[in] MaxSize - Maximum size of the buffer
+ * @retval The The number of characters in the string, including the null character.
+ */
+UINTN
+EFIAPI
+GetInsecureDeviceStateString (
+  OUT CHAR8  *Buffer,
+  IN UINTN   MaxSize
+  )
+{
+  CHAR8         *String;
+  DEVICE_STATE  BitmaskFromIndex;
+  DEVICE_STATE  CurrentDeviceState         = GetDeviceState ();
+  DEVICE_STATE  InsecureDeviceStateSetting = GetInsecureDeviceStateSetting ();
+
+  UINTN  NumberOfDeviceStates = sizeof (DEVICE_STATE_STRINGS) / sizeof (DEVICE_STATE_STRINGS[0]);
+
+  //
+  // Check standard device states
+  //
+  for (UINTN i = 0; i < NumberOfDeviceStates; i++) {
+    BitmaskFromIndex = (1 << (i));
+    if ((DeviceStateBitmaskIsSet (InsecureDeviceStateSetting, BitmaskFromIndex)) &&
+        (DeviceStateBitmaskIsSet (CurrentDeviceState, BitmaskFromIndex)))
+    {
+      //
+      // An insecure device state is detected. Append to the insecure device state string.
+      //
+      String = DEVICE_STATE_STRINGS[i];
+      AsciiStrnCatS (Buffer, MaxSize, String, AsciiStrLen (String));
+    }
+  }
+
+  //
+  // Check OEM defined insecure device states
+  //
+  for (UINTN i = DEVICE_STATE_PLATFORM_MODE_INDEX_START; i <= DEVICE_STATE_PLATFORM_MODE_INDEX_END; i++) {
+    BitmaskFromIndex = (1 << (i));
+    if ((DeviceStateBitmaskIsSet (InsecureDeviceStateSetting, BitmaskFromIndex)) &&
+        (DeviceStateBitmaskIsSet (CurrentDeviceState, BitmaskFromIndex)))

```

Comment:
This API doesn't seem to add value. We can drop it and just directly do the bitwise & here.

Reply:
I added it for readability, specifically for consumption in Tcg2Pei. (https://github.com/microsoft/mu_tiano_plus/pull/342/files#diff-dd252bb953ec11695bb497d99150fe3c48e069945f08a98630c4a2e47baef3b3R1169)
But I can remove it if it's not achieving that goal/it's not much better than bitwise operation. I can add to the function descriptions that use it instead

Reply:
dropped